core/src/main/scala/org/apache/spark/sql/delta/schema/SchemaUtils.scala

```diff
@@ -538,57 +538,53 @@ object SchemaUtils extends DeltaLogging {
-      resolver: Resolver = DELTA_COL_RESOLVER): (Seq[Int], Int) = {
-    def find(column: Seq[String], schema: StructType, stack: Seq[String]): (Seq[Int], Int) = {
-      if (column.isEmpty) return (Nil, schema.size)
-      val thisCol = column.head
-      lazy val columnPath = UnresolvedAttribute(stack :+ thisCol).name
-      val pos = schema.indexWhere(f => resolver(f.name, thisCol))
-      if (pos == -1) {
-        throw new IndexOutOfBoundsException(columnPath)
-      }
-      val colTail = column.tail
-      val (children, lastSize) = (colTail, schema(pos).dataType) match {
-        case (_, s: StructType) =>
-          find(colTail, s, stack :+ thisCol)
-        case (Seq("element", _ @ _*), ArrayType(s: StructType, _)) =>
-          val (child, size) = find(colTail.tail, s, stack :+ thisCol)
-          (ARRAY_ELEMENT_INDEX +: child, size)
-        case (Seq(), ArrayType(s: StructType, _)) =>
-          find(colTail, s, stack :+ thisCol)
-        case (Seq(), ArrayType(_, _)) =>
-          (Nil, 0)
-        case (_, ArrayType(_, _)) =>
-          throw DeltaErrors.incorrectArrayAccessByName(
-            prettyFieldName(stack ++ Seq(thisCol, "element")),
-            prettyFieldName(stack ++ Seq(thisCol)))
-        case (Seq(), MapType(_, _, _)) =>
-          (Nil, 2)
-        case (Seq("key", _ @ _*), MapType(keyType: StructType, _, _)) =>
-          val (child, size) = find(colTail.tail, keyType, stack :+ thisCol)
-          (MAP_KEY_INDEX +: child, size)
-        case (Seq("key"), MapType(_, _, _)) =>
-          (Seq(MAP_KEY_INDEX), 0)
-        case (Seq("value", _ @ _*), MapType(_, valueType: StructType, _)) =>
-          val (child, size) = find(colTail.tail, valueType, stack :+ thisCol)
-          (MAP_VALUE_INDEX +: child, size)
-        case (Seq("value"), MapType(_, _, _)) =>
-          (Seq(MAP_VALUE_INDEX), 0)
-        case (_, MapType(_, _, _)) =>
-          throw DeltaErrors.foundMapTypeColumnException(
-            prettyFieldName(stack ++ Seq(thisCol, "key")),
-            prettyFieldName(stack ++ Seq(thisCol, "value")))
-        case (_, o) =>
-          if (column.length > 1) {
-            throw DeltaErrors.columnPathNotNested(columnPath, o, column)
+      resolver: Resolver = DELTA_COL_RESOLVER): Seq[Int] = {
+    def find(
+        searchPath: Seq[String],
+        currentType: DataType,
+        currentPath: Seq[String] = Nil): Seq[Int] = {
+      if (searchPath.isEmpty) return Nil
+
+      val currentFieldName = searchPath.head
+      val path = currentPath :+ currentFieldName
+      (currentType, currentFieldName) match {
+        case (struct: StructType, _) =>
+          lazy val columnPath = UnresolvedAttribute(path).name
+          val pos = struct.indexWhere(f => resolver(f.name, currentFieldName))
+          if (pos == -1) {
+            throw new IndexOutOfBoundsException(columnPath)
           }
-          (Nil, 0)
+          val childPath = find(searchPath.tail, struct(pos).dataType, path)
```

Comment:
childPosition then? AFAICT we use path for string paths and position for index paths. (Here and below.)